### ISSUE TITLE
Elastic Search

### ISSUE THREAD

**Issue by wawiwa** [bluedenim/log4j-s3-search#90]
Before making a pull request I thought I would ask if there's already full support for elastic search implemented in appender version 2.6? I tried to enabling the elastic search appender in log4j2.xml but it's ignoring my xml settings and using predefined default values. It does however seem to respect only the elasticsearchHosts element. The other three properties: elasticsearchCluster, elasticsearchIndex, elasticsearchType are always set to - type=log , index=logindex , cluster=elasticsearch. Also I noticed the elastic search library is 5.4.3 which is pretty much deprecated, I think it's up to 7.11 now. As a follow up I was able to get the appender working using es transport client 6.8 after 'overidding' a couple of classes and making some changes to set these properties in the Log4j2AppenderBuilder.java. Please let me  know if you would like me to send a PR with updates. We also have a requirement to format the message as json in elasticsearch . I added some code for that as well, but my guess is that we would need another configuration property to indicate that the message is json. Thank you for a really fantastic library.

**Comment by bluedenim** [comment]
Thanks for your interest in the project and bringing this to my attention.

It's been a while since I touched the Elasticsearch (or Solr for that matter) support for the library. I just took a look and, although the config class (ElasticsearchConfiguration) have the fields, looks like the Log4j2 init code (i.e. Log4j2AppenderBuilder's createCachePublisher and getElasticsearchConfigIfEnabled) are not setting the correct fields other than, as you pointed out, the hosts. Ironically, the Log4j 1.x appender seems to have better support for those than the Log4j 2.x appender.

This is a relatively simple bug to fix, I believe. However, it'll take a bit of time for me to spin up a local ES to test.

**Comment by bluedenim** [comment]
The missing config parameters is a bug whereas the JSON message seems like a feature, so I'm leaning to treating them to be two different tasks. 

I'd be happy to see and learn what you have done. Thanks again.

**Comment by wawiwa** [comment]
Thank you for your quick reply.  I saw the summary of changes in the elastic search config PR and it looks great. For the JSON message it does seem a separate task is better suited to track it. Maybe it's a bit of an oddball requirement to have this sort of functionality but we're pretty much locked in to storing it as json instead of an escaped string in elastic search. Here's the snippet of the ElasticsearchPublishHelper that converts the string to json. It just blindly assumes we're using json for the message but for the production code I was thinking if we had  a boolean configuration parameter like elasticsearchMessageAsJson we could use it as a flag to indicate whether or not to parse the message

public void publish(PublishContext context, int sequence, Event event) {
        **try (XContentParser parser = JsonXContent.jsonXContent
                .createParser(NamedXContentRegistry.EMPTY, null, event.getMessage())){**
            String id = String.format("%s-%s-%016d", context.getCacheName(),
                context.getHostName(), offset); 
            XContentBuilder contentBuilder = jsonBuilder()
                .startObject()
                .field("timestamp", timeStamp)
                .field("type", event.getType())
                .field("hostname", context.getHostName())
                .field("offset", offset)
                .field("thread_name", event.getThreadName())
                .field("logger", event.getSource())
                **.field("message")
                .copyCurrentStructure(parser)**
                .array("tags", context.getTags())
                .endObject();
            builder.add(transport.prepareIndex(configuration.getIndex(), configuration.getType(), id)
                .setSource(contentBuilder));
            offset++;
        } catch (Exception ex) {
            System.err.println(String.format("Cannot publish event: %s", ex.getMessage()));
        }
    }

**Comment by bluedenim** [comment]
OK. Thanks. I'll keep this around for reference.

The PR is broken right now because the newer Java client has breaking changes, so I'm trying to figure out how the new client works. I have something that IntelliJ is happy about for now, at least. I'll update the PR when I get a chance.

**Comment by wawiwa** [comment]
While trying to upgrade from the transport client to the rest client I ran into an issue with the new BulkRequest. For some reason its add method validates that the IndexRequest object has a "type" parameter although the type parameter is annotated as deprecated. So not knowing what to do there, I kept it and made it "_doc" for now. The code is working and persisting to ES ok. A little separately , we also had to add an additional field called as "_class" to accommodate spring boot's JPA compatibility. I realize this is more of a one off feature where we can add  custom fields to the Event.class
Here's a snippet of the xml config we're using and the publish() method
<Log4j2Appender name="UserActivityESapndr">
			<PatternLayout pattern="%m%n" />
			<verbose>false</verbose>
			<stagingBufferAge>1</stagingBufferAge>
			<elasticsearchCluster>elasticsearch</elasticsearchCluster>
			<elasticsearchIndex>user-justifications</elasticsearchIndex>
			<elasticsearchType>com.company.entity.UserActivityLogBase</elasticsearchType>
			<elasticsearchHosts>localhost:9200</elasticsearchHosts>
		</Log4j2Appender>

`@Override
    public void publish(PublishContext context, int sequence, Event event) {
        try (XContentParser parser = JsonXContent.jsonXContent
                .createParser(NamedXContentRegistry.EMPTY, null, event.getMessage())){
            String id = String.format("%s-%s-%016d", context.getCacheName(),
                context.getHostName(), offset); 
            XContentBuilder contentBuilder = jsonBuilder()
                .startObject()
                .field("timestamp", timeStamp)
                .field("type", event.getType())
                .field("hostname", context.getHostName())
                .field("offset", offset)
                .field("thread_name", event.getThreadName())
                .field("logger", event.getSource())
                .field("_class",configuration.getType())
                .field("message")
                .copyCurrentStructure(parser)
                .array("tags", context.getTags())
                .endObject();
            // TODO: Why is type validated in elastic search bulk request if it's not required in version 7+ ?
            bulkRequest.add(new IndexRequest(configuration.getIndex()).type("_doc").id(id).source(contentBuilder));
            offset++;
        } catch (Exception ex) {
            System.err.println(String.format("Cannot publish event: %s", ex.getMessage()));
        }
    }`

**Comment by bluedenim** [comment]
Upgrading the client is going to take more work than I expected. Instead, I'm going to fix the bug while staying on the 5.6.x version (latest is 5.6.16) for now. This will at least fix the bug. I've tested it against ES 6.8.14.

I'll start a new branch to get this working with the newer ES (7.10+) versions.

**Comment by bluedenim** [comment]
After #96, you'll be able to add your own implementation of the publish helper to use to publish into Elasticsearch. With that, you can now override the `start`, `publish`, and/or `end `methods to interact with Elasticsearch the way you need.

There is an example of that now in the class https://github.com/bluedenim/log4j-s3-search-samples/blob/master/appender-log4j2-sample/src/main/java/com/van/example/MyElasticsearchPublishHelper.java, configured in via https://github.com/bluedenim/log4j-s3-search-samples/blob/master/appender-log4j2-sample/src/main/resources/log4j2.xml#L64

**Comment by wawiwa** [comment]
Hi Van , that's a really great idea. That makes it very easy for us to implement our changes. As an aside I asked that we complete our PR and it looks like we can start that effort sometime April. It's in our hands once #96 is in place but here's what we have lined up so far 1) implement client 7.6.2 2) add custom field array 3) resolve bulk request deprecation. 4) resolve duplicate hostname in doc id. Thanks again for all of your considerations along the way.

**Comment by bluedenim** [comment]
OK. Release 3.1.0 has the changes.
